docs/specs/priority-queue.md

```diff
@@ -1,70 +1,99 @@
+Whenever such a change to an item happens, the priority queue just be informed in a single call of the list of ALL items whose priority has changed, to ensure that no data-structure invariants are violated when the update-priority algorithmm is executed.
+
+## The proposed classes
+
+There are actually two PriorityQueue classes, one to support each scenario. PriorityQueue<TItem, TPriority> supports the *extrinsic priority* scenario, and PrioritySet<T> supports the *intrinsic priority* scenario. 
+
+## `PrioritySet<T>` (for intrinsic priority)
+
+`PrioritySet<T>` is a set of items with intrinsic priorities. It must be initialized with a 'priority comparer' which understands how to compare the intrinsic priorities of each item.
+
+It differs from `SortedSet<T>` because it allows separate items to have equal priorities, and because it does not maintain strict ordering at all times, and because it allows that priorities are mutable.
+
+If you provide `DefaultComparer<T>` as your priority comparer, and never mutate priorities, you will get similar behavior to `SortedSet<T>`, where priorities are the items and priorities are all distinct.
 
 ```csharp
-public class PriorityQueue<T> : IQueue<T>
+public class PrioritySet<T> : ISet<T>, ICollection<T>, IEnumerable<T>, IEnumerable, 
 {
-    public PriorityQueue();
-    public PriorityQueue(IComparer<T> comparer);
-    public PriorityQueue(IEnumerable<T> collection);
-    public PriorityQueue(IEnumerable<T> collection, IComparer<T> comparer);
+    public PriorityQueue(IComparer<T> priorityComparer);
+    public PriorityQueue(IComparer<T> priorityComparer, IEqualityComparer<T> equalityComparer);
+
+    public PriorityQueue(IEnumerable<T> collection, IComparer<T> priorityComparer);
+    public PriorityQueue(IEnumerable<T> collection, IComparer<T> priorityComparer, IEqualityComparer<T> equalityComparer);
+
+    public IComparer<T> PriorityComparer { get; } // compares items using .CompareTo() to see which is higher priority
+    public IComparer<T> EqualityComparer { get; } // comapres items using .GetHashCode() and .Equals() to see if they are the same as existing items in the collection
 
-    public IComparer<T> Comparer { get; }
     public int Count { get; }
+    public bool IsReadOnly { get; }
 
-    public bool IsEmpty { get; }
+    public bool Add(T item);
     public void Clear();
-    public bool Contains(T element);
-
-    public void Enqueue(T element);
+    public bool Contains(T item);
+    public void CopyTo(T[] item, int index);
+    public bool Remove(T item);
 
     public T Peek();
     public T Dequeue();
-    public bool Remove(T element);
 
-    public bool TryPeek(out T element);
-    public bool TryDequeue(out T element);
+    public bool TryPeek(out T item);
+    public bool TryDequeue(out T item);
+
+    public void NotifyPriorityChanged(T item);
```

Comment:
not sure what the `NotifyPriorityChanged` methods are for?

Reply:
@nietras It is the simplest way I could think for PrioritySet is to support elements whose priority mutates over time.
I am still conflicted on whether it is a good feature, as it can let in consistency bugs - maybe it is better to force people to model mutable priority scenarios using PriorityQueue and not PrioritySet...

Reply:
@TimLovellSmith I guess my main objection here is the name, it is really poor in my view. It is not notifying anything. If anything it sounds more like you want it to do an `Update` but how would that work if the item itself is both priority and key? How can it then update the old item? So this kind of implicitly assumes the `item` has **two** properties a key and a priority, which does not seem to be a good assumption.

If that is I understand the purpose here. It is like the `priorityQueue[key] = priority` just with `item` being both right? So basically changing an item should perhaps involve first deleting the old and then adding new? Or with a proper `Replace(T old, T new)` method.